src/config/config-factory.ts

```diff
@@ -1,24 +1,107 @@
 import maliConfig from './chis-ml';
+import path from 'path';
+import fs from 'fs';
 
-export const CONFIG_MAP: { [key: string]: PartnerConfig } = {
-  'CHIS-KE': kenyaConfig,
-  'CHIS-TG': togoConfig,
-  'CHIS-CIV': civConfig,
-  'CHIS-ML': maliConfig
-};
-
-export default function getConfigByKey(key: string = 'CHIS-KE'): PartnerConfig {
-  const usingKey = key.toUpperCase();
-  console.log(`Using configuration: ${key}`);
-  const result = CONFIG_MAP[usingKey];
-  if (!result) {
-    const available = JSON.stringify(Object.keys(CONFIG_MAP));
-    throw Error(`Failed to start: Cannot find configuration "${usingKey}". Configurations available are ${available}`);
+export default class ConfigFactory<T = Record<string, any>> {
+  private config: PartnerConfig | null = null;
+  private static configFactoryInstance: ConfigFactory;
+  private filePath: string;
+
+  private DEFAULT_CONFIG_MAP: { [key: string]: PartnerConfig } = {
+    'CHIS-KE': kenyaConfig,
+    'CHIS-TG': togoConfig,
+    'CHIS-CIV': civConfig,
+    'CHIS-ML': maliConfig
+  };
+
+  constructor() {
+    this.filePath = path.join(this.getConfigUploadDirectory(), 'config.json');
+  }
+
+  static getConfigFactory<T = Record<string, any>>(): ConfigFactory<T> {
+    if (!ConfigFactory.configFactoryInstance) {
+      ConfigFactory.configFactoryInstance = new ConfigFactory();
+    }
+
+    return ConfigFactory.configFactoryInstance as ConfigFactory<T>;
   }
 
-  return result;
+  loadConfig(key?: string): PartnerConfig {
+
+    if (this.config) {
+      return this.config;
+    }
+
+    if (fs.existsSync(this.filePath)) {
+      this.config = this.readConfig();
+      console.log(`Using uploaded configuration: ${this.filePath}`);
+      return this.config;
+    }
+
+    //set default config
+    if (!key) {
+      console.log('no configuration found. Setting to default');
+      const defaultConfig = kenyaConfig;
+      defaultConfig.config.domains = [{
+        friendly: 'Localhost',
+        domain: 'localhost'
+      }];
+      return this.config = defaultConfig;
+    }
+
+    const usingKey = key.toUpperCase();
+    console.log(`Using configuration: ${key}`);
+    const result = this.DEFAULT_CONFIG_MAP[usingKey];
+    if (!result) {
+      const available = JSON.stringify(Object.keys(this.DEFAULT_CONFIG_MAP));
+      throw Error(`Failed to start: Cannot find configuration "${usingKey}". Configurations available are ${available}`);
+    }
+
+    this.config = result;
+    return this.config;
+
+  }
+
+  refreshConfig(): PartnerConfig | null {
+    if (fs.existsSync(this.filePath)) {
+      this.config = this.readConfig();
+      console.log(`Using uploaded configuration: ${this.filePath}`);
+      return this.config;
+    }
+    return null;
+  }
+
+  writeConfig(jsonConfigData: ConfigSystem): void {
+    try {
+      const jsonString: string = JSON.stringify(jsonConfigData, null, 2);
+      fs.writeFileSync(this.filePath, jsonString);
+      this.refreshConfig();
```

Comment:
since `refreshConfig` is only used here, maybe you can just directly set the config here after writing it. Saves you a trip to the filesystem. 

nitpick: the parameter name `jsonConfigData` is kinda confusing, maybe just `config` is fine?

Reply:
lets leave it there keeps it organized the writeconfig will be doing more than it describes 
 besides its a great action to have on the class